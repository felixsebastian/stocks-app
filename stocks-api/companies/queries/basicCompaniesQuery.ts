import db from "../../db";

export default (exchange: string, scoreFilter: string) => {
  const query = db
    .with("lastPrices", (qb) =>
      qb
        .select("company_id", "price AS last_price", "date AS last_price_date")
        .from("swsCompanyPriceClose")
        .groupBy("company_id")
        .orderBy("date_created"),
    )
    .select(
      "swsCompany.id",
      "name",
      "ticker_symbol",
      "exchange_country_iso",
      "exchange_symbol",
      "last_price",
      "last_price_date",
      "total AS snowflake_score",
      db.raw("COUNT(*) OVER() AS total_companies"),
    )
    .from("swsCompany")
    .leftJoin("swsCompanyScore", "swsCompany.id", "swsCompanyScore.company_id");

  if (exchange !== "none") query.where("exchange_symbol", "=", exchange);

  if (scoreFilter !== "none") {
    query.where("snowflake_score", ">", parseInt(scoreFilter.slice(3), 10));
  }

  return query;
};
