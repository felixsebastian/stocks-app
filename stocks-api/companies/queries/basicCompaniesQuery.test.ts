import basicCompaniesQuery from "./basicCompaniesQuery";

describe("basicCompaniesQuery", () => {
  it("should include the exchange filter", () => {
    expect(
      basicCompaniesQuery("some exchange", "none").toString(),
    ).toMatchSnapshot();
  });

  it("should include the score filter", () => {
    expect(
      basicCompaniesQuery("none", "some score filter").toString(),
    ).toMatchSnapshot();
  });
});
