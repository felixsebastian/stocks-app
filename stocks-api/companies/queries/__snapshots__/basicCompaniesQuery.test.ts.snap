// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basicCompaniesQuery should include the exchange filter 1`] = `"with \`lastPrices\` as (select \`company_id\`, \`price\` as \`last_price\`, \`date\` as \`last_price_date\` from \`swsCompanyPriceClose\` group by \`company_id\` order by \`date_created\` asc) select \`swsCompany\`.\`id\`, \`name\`, \`ticker_symbol\`, \`exchange_country_iso\`, \`exchange_symbol\`, \`last_price\`, \`last_price_date\`, \`total\` as \`snowflake_score\`, COUNT(*) OVER() AS total_companies from \`swsCompany\` left join \`swsCompanyScore\` on \`swsCompany\`.\`id\` = \`swsCompanyScore\`.\`company_id\` left join \`lastPrices\` on \`swsCompany\`.\`id\` = \`lastPrices\`.\`company_id\` where \`exchange_symbol\` = 'some exchange'"`;

exports[`basicCompaniesQuery should include the score filter 1`] = `"with \`lastPrices\` as (select \`company_id\`, \`price\` as \`last_price\`, \`date\` as \`last_price_date\` from \`swsCompanyPriceClose\` group by \`company_id\` order by \`date_created\` asc) select \`swsCompany\`.\`id\`, \`name\`, \`ticker_symbol\`, \`exchange_country_iso\`, \`exchange_symbol\`, \`last_price\`, \`last_price_date\`, \`total\` as \`snowflake_score\`, COUNT(*) OVER() AS total_companies from \`swsCompany\` left join \`swsCompanyScore\` on \`swsCompany\`.\`id\` = \`swsCompanyScore\`.\`company_id\` left join \`lastPrices\` on \`swsCompany\`.\`id\` = \`lastPrices\`.\`company_id\` where \`snowflake_score\` > NaN"`;

exports[`should change when exchange filter is provided 1`] = `"with \`lastPrices\` as (select \`company_id\`, \`price\` as \`last_price\`, \`date\` as \`last_price_date\` from \`swsCompanyPriceClose\` group by \`company_id\` order by \`date_created\` asc) select \`swsCompany\`.\`id\`, \`name\`, \`ticker_symbol\`, \`exchange_country_iso\`, \`exchange_symbol\`, \`last_price\`, \`last_price_date\`, \`total\` as \`snowflake_score\`, COUNT(*) OVER() AS total_companies from \`swsCompany\` left join \`swsCompanyScore\` on \`swsCompany\`.\`id\` = \`swsCompanyScore\`.\`company_id\` left join \`lastPrices\` on \`swsCompany\`.\`id\` = \`lastPrices\`.\`company_id\` where \`exchange_symbol\` = 'some exchange'"`;

exports[`should change when score filter is provided 1`] = `"with \`lastPrices\` as (select \`company_id\`, \`price\` as \`last_price\`, \`date\` as \`last_price_date\` from \`swsCompanyPriceClose\` group by \`company_id\` order by \`date_created\` asc) select \`swsCompany\`.\`id\`, \`name\`, \`ticker_symbol\`, \`exchange_country_iso\`, \`exchange_symbol\`, \`last_price\`, \`last_price_date\`, \`total\` as \`snowflake_score\`, COUNT(*) OVER() AS total_companies from \`swsCompany\` left join \`swsCompanyScore\` on \`swsCompany\`.\`id\` = \`swsCompanyScore\`.\`company_id\` left join \`lastPrices\` on \`swsCompany\`.\`id\` = \`lastPrices\`.\`company_id\` where \`snowflake_score\` > NaN"`;
